PEP 771: Amendments following discussion (#4428)
python/peps: peps/pep-0771.rst

* `glueviz <https://www.glueviz.org>`_: ``glueviz[qt]``

In all three cases, installing the package without any extras results in a

> broken installation, and this is a commonly reported issue for some of these

Reviewer: ```suggestion
broken installation, and this is a commonly reported support issue for some of these
```

packages.

Rationale
=========

> A number of possible solutions have been extensively discussed by

Reviewer 2: I noticed that too, but that's on me.  PEBKAC!  Apologies, and thanks for applying them manually.

then the default extra should be installed because ``package`` appears at least
once with no extras specified.

An empty set of extras, such as ``package[]`` should be interpreted as meaning
that the package should be installed *without* any default extras (unless

> ``package`` appears elsewhere in the dependency tree as mentioned above), so
> would provide a universal way of obtaining a minimal installation of a package.

Reviewer: ```suggestion
``package`` appears elsewhere in the dependency tree as mentioned above). This
would provide a universal way of obtaining a minimal installation of a package.
```

    ...

For tools that behave like this (rather than raising an error), if an extra is
recognized as invalid in a dependency specification, it should be ignored, and
if all specified extras are invalid, then this should be considered equivalent
to ``package[]`` (rather than ``package``) and *not* install any default extras.

Finally, we note (as also discussed in `Relying on tooling to deselect
any default extras`_) that package installers are allowed to implement their
own options to control the above behavior, for example implementing an option
that disables default extras for some or all packages regardless of where these

> packages appear in the dependency tree -- although if such options are
> implemented, they should be opt-in, and users should experience the above
> behavior as default.

Reviewer: ```suggestion
packages appear in the dependency tree. If such tool-specific options are
implemented, tool developers should make these opt-in,
and users should experience the above PEP 771 behavior as default.
```

    $ pip install napari[pyside6]

Supporting extras that should not remove default extras
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

An additional case we consider here is where a package maintainer wants to
support the ability for users to opt-in to non-default extras, without removing
default extras. Essentially, they might want:

* ``package[]`` to give an installation without any extras
* ``package`` to install recommended dependencies (in a ``recommended`` extras)

> * ``package[alternative]`` to not install default extras, but to install an alternative set of optional dependencies (in an ``alternative`` extras)

Reviewer: Do we anticipate `alternative` and `additional` to be commonly used? If so, it may make sense somewhere in the PEP, even an appendix to provide how that differs from `dev`, `testing`, `docs`, etc. today.

Author: No, ``alternative`` and ``additional`` are not names that I expect people to use - they are just example names to distinguish between extras that add dependencies and ones that replace the default ones.

    $ pip install astropy[recommended, jupyter]

Packages with multiple kinds of defaults
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In some cases, it may be that packages need multiple kinds of defaults. As an example,
in `Packages requiring at least one backend or frontend`_, we considered the case of packages
that have *either* backends or frontends, but in some cases, packages may have to support

> backends *and* frontends, and want to specify a default frontend and a default backend.

Reviewer: Or may need multiple backends or frontends. I could see someone having multiple backends available for visualization libraries and giving users access to all through their GUI.

Author: I agree, I've tried to reword this by saying 'one or more default frontend and one or more default backend'

available, so they can be shown at the same time how do properly use the extras for
backends and frontends.

One option to increase user friendliness is that maintainers can create extras
called for example ``defaultbackend`` and ``defaultfrontend`` which do install
the default backend and frontend. They can then recommend that users do:

.. code-block:: console

    $ pip install package  # installs default backend and frontend
    $ pip install package[]  # installs no backends or frontends
    $ pip install package[backend1, defaultfrontend]  # installs backend1 and default frontend
    $ pip install package[defaultbackend, frontend2]  # installs frontend2 and default backend
    $ pip install package[backend1, frontend2]  # installs backend1 and frontend2

This would allow (if desired) users to then get whatever the recommended backend
is, even if that default changes in time.

If there is a desire to implement a better solution in future, we believe this
PEP should not preclude this. For example, one could imagine in future adding the
ability for an extra to specify *which* default extras it disables rather than

> disabling all default extras, but the default could still be for explicitly

Author: Wow yes that was a lot of 'defaults' - reworded

may be changes in behavior in packaging tools due to the change in the
assumption about what impact an extra can have.

It is worth noting that the recently-accepted :pep:`751` defines a new file
format which is intended to replace alternatives such as the ``pip freeze``
output and other tools in future. The new file format is designed so that the
packages in the file are installed *without* resolving dependencies, which means
that it will be fully compatible with default extras as specified in this PEP,

> and will avoid the issue with ``pip freeze``/``pip install -r`` mentioned above.

Author: Yes that's definitely an option though I think some people had objections to this approach previously - in any case this could be discussed in the next DPO thread, but my feeling is that it isn't needed since PEP 751 is happening anyway, regardless of PEP 771

> and will avoid the issue with ``pip freeze``/``pip install -r`` mentioned above.

Reviewer 3: `pip freeze` is blatantly in violation of this proposed PEP, if it does *not* include `[]` for at least any packages that include default extras.

Whether it does so for all packages or just the ones that have default extras is a development tradeoff between slightly uglier files (more `[]` than are necessary) and more work spent to analyze the metadata of installed packages to determine whether there are `Default-Extra` fields to suppress.

Of course "pip freeze is deprecated so it will just not follow the PEP" is a potential decision to make but that should be explicitly spelled out here. You went to quite some effort to write this section of the backwards compatibility documentation, why stop halfway through? :)

> and will avoid the issue with ``pip freeze``/``pip install -r`` mentioned above.

Author: @warsawnv to be clear, the new lock file format from PEP 751 won't have a need for [] even for packages with default extras since that lock file should be interpreted without resolving dependencies.

> and will avoid the issue with ``pip freeze``/``pip install -r`` mentioned above.

Author: In any case, the suggestion of including [] in pip freeze output is probably something that should be discussed in the next DPO thread rather than discuss it extensively here?

  the extra explicitly as long as possible (until it is clear that most/all users
  are using package installers that implement this PEP). There is no downside to
  keeping the extra be explicitly mentioned, but this will ensure that users with
  modern tooling who do not read documentation (which may be a non-negligible
  fraction of the user community) will start getting the recommended
  dependencies by default.

* Since prior to this PEP, ``package[]`` was equivalent to ``package``,
  authors will be able to document ``package[]`` as a backward-compatible
  universal way of getting a minimal installation. For packages that define

>   default extras, installing ``package[]`` will always give a minimal

Author: Yes ``package[]`` will continue to be allowed for packages that don't use default extras, and I've now added a sentence to make this explicit.

mean that all extras need to be defaults, and there is still scope for users to
explicitly opt in to non-default extras.

In essence, default extras should be considered just as 'weighty' as a required
dependency. In some cases, if a package is widely used by many others, if it
adds a default extra, then unless all downstream packages update their
dependencies to specifically request a minimal installation, the defaults will
often end up getting installed. A concrete example of this is that the `pytest
<https://docs.pytest.org>`_ package has almost 1500 plugins at the time of
writing, all of which depend on pytest. If pytest was to start using a default
extra, it would need to consider this very carefully, as if the plugins are not
updated, installing any plugin will result in the default extra being
installed. This is not to say that it should not use default extras, but just

> that the decision should be considered and planned carefully.

Reviewer: You lost me a bit in this example. I'm going to parrot back to see if I understood this:
- Be kind to downstream packages by not installing too many default extras
- If I were to install pytest as a required dependency to say scipy, then all the default extras for pytest (1500 plugins) would be installed when scipy is installed.

Reviewer 4: Your understanding looks correct to me!

---

Here's my attempt at rephrasing this section to be a bit clearer (breaking up some of the long sentences and tightening up the language).

```suggestion
Default extras should generally be treated with the same "weight" as required dependencies. When a package is widely used, introducing a default extra will result in that extra's dependencies being transitively included -- unless all downstream packages are updated to explicitly opt out using minimal installation specifications.

As an example, the `pytest <https://docs.pytest.org/>`_ package currently has nearly 1,500 plugins that depend on it. If pytest were to add a default extra and those plugins were not updated accordingly, installing a plugin would include the default extras' dependencies. This doesn’t preclude the use of default extras, but addition of default extras requires careful evaluation of its downstream effects.
```

Author: Great rephrasing - thank you!

mean that all extras need to be defaults, and there is still scope for users to
explicitly opt in to non-default extras.

In essence, default extras should be considered just as 'weighty' as a required
dependency. In some cases, if a package is widely used by many others, if it
adds a default extra, then unless all downstream packages update their
dependencies to specifically request a minimal installation, the defaults will

> Default extras should generally be treated with the same "weight" as required dependencies. When a package is widely used, introducing a default extra will result in that extra's dependencies being transitively included -- unless all downstream packages are updated to explicitly opt out using minimal installation specifications.

Reviewer 2: Something seems broken here!

    └── package2
        └── package1

If ``package1`` has a default extra named ``recommended`` then:

.. code-block:: console

    $ pip install package1[]

will still result in the ``recommended`` extra being installed if ``package2``
continues to depend on ``package1`` (with no extras specified). This could be
solved by changing the dependency tree to instead be::

    package1
    └── package2
        └── package1[]

> assuming that ``package2`` indeed does not depend on any features provided by

Reviewer 2: ```suggestion
assuming that ``package2`` indeed does not depend on any features provided by
```
assuming that indeed ``package2`` does not depend on any features provided by

    └── package2
        └── package1

If ``package1`` has a default extra named ``recommended`` then:

.. code-block:: console

    $ pip install package1[]

will still result in the ``recommended`` extra being installed if ``package2``
continues to depend on ``package1`` (with no extras specified). This could be
solved by changing the dependency tree to instead be::

    package1
    └── package2
        └── package1[]

assuming that ``package2`` indeed does not depend on any features provided by
the extra dependencies of ``package1``. Authors therefore need to carefully
consider a migration plan, coordinating with the authors of ``package2``.

Documenting packages with default extras
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Regardless of how default extras are used, package authors should aim to ensure
> that the documentation for their package makes it clear how extras are to be
> used. 'Best practices' documentation should mention:

Reviewer: ```suggestion
Regardless of how default extras are used, package authors should aim to ensure
that their package's documentation makes it clear how extras are to be
used. 'Best practices' documentation should mention:
```
